adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+func getHuaweiAdsReqNetWorkInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	if openRTBRequest.Device != nil {
+		var network network
+		if openRTBRequest.Device.ConnectionType != nil {
+			network.Type = int32(*openRTBRequest.Device.ConnectionType)
+		} else {
+			network.Type = DefaultUnknownNetworkType
+		}
+
+		var cellInfos []cellInfo
+		if openRTBRequest.Device.MCCMNC != "" {
+			var arr = strings.Split(openRTBRequest.Device.MCCMNC, "-")
+			network.Carrier = 0
+			if len(arr) >= 2 {
+				cellInfos = append(cellInfos, cellInfo{
+					Mcc: arr[0],
+					Mnc: arr[1],
+				})
+				var str = arr[0] + arr[1]
+				if str == "46000" || str == "46002" || str == "46007" {
+					network.Carrier = 2
+				} else if str == "46001" || str == "46006" {
+					network.Carrier = 1
+				} else if str == "46003" || str == "46005" || str == "46011" {
+					network.Carrier = 3
+				} else {
+					network.Carrier = 99
+				}
+			}
+		}
+		network.CellInfo = cellInfos
+		request.Network = network
+	}
+}
+
+// get regs information for HuaweiAds request
+func getHuaweiAdsReqRegsInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	if openRTBRequest.Regs != nil && openRTBRequest.Regs.COPPA >= 0 {
+		var regs regs
+		regs.Coppa = int32(openRTBRequest.Regs.COPPA)
+		request.Regs = regs
+	}
+}
+
+// get geo information for HuaweiAds request
+func getHuaweiAdsReqGeoInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	if openRTBRequest.Device != nil && openRTBRequest.Device.Geo != nil {
+		var geo geo
+		geo.Lon = float32(openRTBRequest.Device.Geo.Lon)
+		geo.Lat = float32(openRTBRequest.Device.Geo.Lat)
+		geo.Accuracy = int32(openRTBRequest.Device.Geo.Accuracy)
+		geo.Lastfix = int32(openRTBRequest.Device.Geo.LastFix)
+		request.Geo = geo
+	}
+}
+
+func unmarshalExtImpHuaweiAds(openRTBImp *openrtb2.Imp) (*openrtb_ext.ExtImpHuaweiAds, error) {
+	var bidderExt adapters.ExtImpBidder
+	var huaweiAdsImpExt openrtb_ext.ExtImpHuaweiAds
+	if err := json.Unmarshal(openRTBImp.Ext, &bidderExt); err != nil {
+		return nil, errors.New("Unmarshal: openRTBImp.Ext -> bidderExt failed")
+	}
+	if err := json.Unmarshal(bidderExt.Bidder, &huaweiAdsImpExt); err != nil {
+		return nil, errors.New("Unmarshal: bidderExt.Bidder -> huaweiAdsImpExt failed")
+	}
+	if huaweiAdsImpExt.SlotId == "" {
+		return nil, errors.New("ExtImpHuaweiAds: slotid is empty.")
+	}
+	if huaweiAdsImpExt.Adtype == "" {
+		return nil, errors.New("ExtImpHuaweiAds: adtype is empty.")
+	}
+	if huaweiAdsImpExt.PublisherId == "" {
+		return nil, errors.New("ExtHuaweiAds: publisherid is empty.")
+	}
+	if huaweiAdsImpExt.SignKey == "" {
+		return nil, errors.New("ExtHuaweiAds: signkey is empty.")
+	}
+	if huaweiAdsImpExt.KeyId == "" {
+		return nil, errors.New("ExtImpHuaweiAds: keyid is empty.")
+	}
+	return &huaweiAdsImpExt, nil
+}
+
+// check response status code
+func checkRespStatusCode(response *adapters.ResponseData) error {
```

Comment:
Please add a json test file for each condition in the `checkRespStatusCode` method.

Reply:
done